### PR TITLE
Validate name in AgentPluginManifest

### LINKED ISSUE #2735: Limit allowed characters in  `AgentPluginManifest.name`
# Refactor

## Component(s) to be refactored

- AgentPluginManifest

## Explanation

Agent plugin names are used to create directories. Plugin names containing certain characters, such as "/" or "..",  could lead to directory traversal vulnerabilities. While the code that extracts plugins mitigates this, a defense-in-depth approach is prudent.

Limit the characters in the plugin name to `[a-zA-Z0-9_-]`

### PR DESCRIPTION
**Pull request by shreyamalviya**
# What does this PR do?

Fixes #2735 

## PR Checklist
* [x] Have you added an explanation of what your changes do and why you'd like to include them?
* [x] Is the TravisCI build passing?
* [ ] ~Was the CHANGELOG.md updated to reflect the changes?~
* [ ] ~Was the documentation framework updated to reflect the changes?~
* [x] Have you checked that you haven't introduced any duplicate code?

## Testing Checklist

* [x] Added relevant unit tests?
* [x] Have you successfully tested your changes locally? Elaborate:
    > UTs pass
* [ ] ~If applicable, add screenshots or log transcripts of the feature working~

### CODE REVIEW

File: monkey/common/agent_plugins/agent_plugin_manifest.py [hunk 1]
```diff
@@ -31,3 +34,11 @@ class AgentPluginManifest(InfectionMonkeyBaseModel):
     description: Optional[str]
     link_to_documentation: Optional[str]
     safe: bool = False
+
+    @validator("name")
+    def validate_name(cls, name):
```
Comment (VakarisZ) [inline monkey/common/agent_plugins/agent_plugin_manifest.py:39]: This looks like a reusable filename validator. Maybe add a file `monkey/common/types/filename.py` and add a validator there (named `validate_filename`)?
Then in the `AgentPluginManifest` add
```
_validate_filename = validator('name', allow_reuse=True)(validate_filename)
```

File: monkey/tests/unit_tests/infection_monkey/puppet/test_plugin_source_extractor.py [hunk 2]
```diff
@@ -1,16 +1,44 @@
...
+    name: str
+    plugin_type: AgentPluginType
+    supported_operating_systems: Tuple[OperatingSystem, ...] = (
+        OperatingSystem.WINDOWS,
+        OperatingSystem.LINUX,
+    )
+    title: Optional[str]
+    description: Optional[str]
+    link_to_documentation: Optional[str]
+    safe: bool = False
```
Comment (VakarisZ) [inline monkey/tests/unit_tests/infection_monkey/puppet/test_plugin_source_extractor.py:30]: Maybe patching the validator would be better?

Comment (mssalvatore) [reply monkey/tests/unit_tests/infection_monkey/puppet/test_plugin_source_extractor.py:30]: Can we just inherit `AgentPluginManifest` and override `name`?

File: monkey/common/types/plugin_name.py [hunk 3]
```diff
@@ -0,0 +1,13 @@
...
+
+
+class PluginName(ConstrainedStr):
+    """
+    Plugin name in snake_case
+    """
+
+    strip_whitespace = True
+    to_lower = True
+    regex = re.compile("^[a-zA-Z0-9_]+$")
```
Comment (mssalvatore) [inline monkey/common/types/plugin_name.py:13]: Why are we doing `to_lower` but allowing `A-Z` in the regex? I see no reason to limit this to lowercase characters.

```suggestion
    regex = re.compile("^[a-zA-Z0-9_]+$")
```

**Review by ilija-lazoroski — APPROVED** [review]

**Review by mssalvatore — CHANGES_REQUESTED** [review]

File: monkey/common/types/plugin_name.py (hunk 3 above)
Comment (VakarisZ) [reply monkey/common/types/plugin_name.py:13]: regex is validation, to_lower is processing. What happens if we have two plugins SSH and ssh? Are they different? Maybe it's worth allowing A-Z, but making sure that the name is unique

File: monkey/common/agent_plugins/agent_plugin_manifest.py (hunk 1 above)
Comment (shreyamalviya) [reply monkey/common/agent_plugins/agent_plugin_manifest.py]: Seems unnecessary to me right now. We can move it when (if) we need to reuse it.

**Review by mssalvatore — APPROVED** [review]

**Pull request merged**